subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java

```diff
@@ -122,10 +123,13 @@ protected FileLookup getFileLookup() {
     void generate() {
         File jarFileDestination = getJarFile();
         File unixScript = getScriptFile();
+        File propertiesFile = getPropertiesFile();
         FileResolver resolver = getFileLookup().getFileResolver(unixScript.getParentFile());
         String jarFileRelativePath = resolver.resolveAsRelativePath(jarFileDestination);
+        Properties existingProperties = propertiesFile.exists() ? GUtil.loadProperties(propertiesFile) : null;
 
-        writeProperties(getPropertiesFile());
+        checkProperties(existingProperties);
+        writeProperties(propertiesFile, existingProperties);
```

Comment:
It might be a little simpler to understand what's going on here by generating the new properties first into a structure we can compare before writing them to a file.

We could do something like this:
1. Read the existing properties into `existingProperties`
2. Generate the _new_ properties into `newProperties`
3. Figure out which scenario we're in and if we need to fail. pseudo-code: 
```
// User hasn't specified a new checksum
if (newProperties.checksum == null) {
    // Distribution URL hasn't changed
    if (existingProperties.distributionUrl == newProperties.distributionUrl) { 
        // keep the checksum from the existing properties (may be null)
        newProperties.checksum = existingProperties.checksum
    } else if (existingProperties.checksum!=null) {
        // distribution URL has changed and the existing properties had a checksum
        // fail because the user needs to give us a checksum now
        throw new GradleException(...)
    }
} else {
    // New distribution checksum has been specified, trust whatever the user provided
}
```
4. Write `newProperties` to the properties file

Reply:
Yeah, it sounds good, i've tried to implement this in separate branch and this made a code in Wrapper more clear and simple to understand. Thank you!